### PR TITLE
 add subnet level connectivity output, WIP

### PR DESCRIPTION
**Pull request by adisos**
issue #62 - initial support: consider full subnet ranges only

### CODE REVIEW

File: pkg/vpcmodel/abstractVPC.go [hunk 1]
```diff
@@ -29,6 +39,7 @@ type Node interface {
 	IsPublicInternet() bool
 	Details() string
 	DetailsMap() map[string]string
+	Kind() string
```
Comment (zivnevo) [inline pkg/vpcmodel/abstractVPC.go:42]: Consider moving `Details()`, `DetailsMap()` and `Kind()` to `NamedResourceIntf` (and maybe rename it to `VPCResourceIntf`)

File: pkg/vpcmodel/cloudConfig.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+// connectivity implied from nacl configuration applied on a certain subnet in the vpc
+func (c *CloudConfig) GetConnectivityOutputPerEachSubnetSeparately() string {
+	// iterate over all subnets, collect all outputs per subnet connectivity
+	for _, r := range c.FilterResources {
+		if r.Kind() == NaclLayer {
+			return r.GetConnectivityOutputPerEachElemSeparately()
+		}
+	}
+	return ""
+}
```
Comment (zivnevo) [inline pkg/vpcmodel/cloudConfig.go:45]: Doesn't seem like the right place for this function. Move to `subnetsConnectivity.go`?

File: pkg/ibmvpc/nacl_analysis.go [hunk 3]
```diff
@@ -378,31 +380,70 @@ func (na *NACLAnalyzer) addAnalysisPerSubnet(subnetCidr string) {
...
+		sortedKeys = append(sortedKeys, k)
+	}
+	sort.Strings(sortedKeys)
+
+	for _, k := range sortedKeys {
+		disjointSubnetCidr := k
+		connectivityRes := m[k]
+		strResult += getDisjointSubnetCidrStr(disjointSubnetCidr)
+		strResult += connectivityRes.string()
+	}
```
Comment (zivnevo) [inline pkg/ibmvpc/nacl_analysis.go:413]:
```suggestion
	for _, k := range sortedKeys {
		connectivityRes := m[k]
		strResult += getDisjointSubnetCidrStr(k)
		strResult += connectivityRes.string()
	}
```

**Review by zivnevo — CHANGES_REQUESTED** [review]
Some comments to start with. I didn't review the `pkg/vpcmodel/*Connectivity.go` files yet.

File: cmd/analyzer/parse_args.go [hunk 4]
```diff
@@ -5,24 +5,46 @@ import (
...
 const (
+	// output formats supported
 	JSONFormat = "json"
 	TEXTFormat = "txt"
 	MDFormat   = "md"
+
+	// connectivity analysis types supported
+	VsiLevel     = "vsiLevel"    // vsi to vsi connectivity analysis
+	SubnetsLevel = "subnetLevel" // subnet to subnet connectivity analysis
+	DebugSubnet  = "debugSubnet" // single subnet connectivity analysis
```
[the remark this answers is not in the document]
Comment (adisos) [reply cmd/analyzer/parse_args.go:25]: opened an issue for this item.

File: pkg/ibmvpc/analysis_output_test.go [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+			useCases: []outputUseCase{uc1, uc2, uc3a},
+		},
+		{
+			name:     "demo_with_instances",
+			useCases: []outputUseCase{uc1, uc2, uc3a},
+		},
+	}
+
+	for _, tt := range tests {
+		runTest(t, tt, mode)
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/analysis_output_test.go:113]: applied this in the next PR (branch `connectivity_grouping_issue_74` on top of this branch) .

File: pkg/vpcmodel/output.go [hunk 6]
```diff
@@ -40,6 +40,10 @@ func (o *OutputGenerator) SetOutputFile(outFileName string, f OutFormat) {
 	o.outputFiles[f] = outFileName
 }
 
+func (o *OutputGenerator) GetOutputFile(f OutFormat) string {
+	return o.outputFiles[f]
+}
+
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/output.go:46]: applied this in the next PR (branch connectivity_grouping_issue_74 on top of this branch) .

File: pkg/ibmvpc/examples/acl_testing3_analysisPerSubnetSeparately.txt [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+remote: 0.0.0.0-10.240.19.255, conn: No Connections
+remote: 10.240.20.0-10.240.20.255, conn: All Connections
+remote: 10.240.21.0-10.240.29.255, conn: No Connections
+remote: 10.240.30.0-10.240.30.255, conn: All Connections
+remote: 10.240.31.0-255.255.255.255, conn: No Connections
+egressConnectivity:
+
+local ip-block: 10.240.10.0-10.240.10.255
+remote: 0.0.0.0-10.240.19.255, conn: No Connections
+remote: 10.240.20.0-10.240.20.255, conn: protocol: TCP,UDP
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/examples/acl_testing3_analysisPerSubnetSeparately.txt:13]: Improved this in the next PR (branch connectivity_grouping_issue_74 on top of this branch) .

File: pkg/ibmvpc/examples/demo_with_instances.txt [hunk 8]
```diff
@@ -0,0 +1,1078 @@
+=================================== distributed inbound/outbound connections:
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/examples/demo_with_instances.txt:1]: applied this in the next PR (branch connectivity_grouping_issue_74 on top of this branch) .

File: pkg/ibmvpc/nacl_analysis.go (hunk 3 above)
Comment (adisos) [reply pkg/ibmvpc/nacl_analysis.go:413]: changed this function in the next PR (branch connectivity_grouping_issue_74 on top of this branch).

File: pkg/vpcmodel/abstractVPC.go (hunk 1 above)
Comment (adisos) [reply pkg/vpcmodel/abstractVPC.go:42]: opened an issue to apply this in a separate PR

File: cmd/analyzer/parse_args.go [hunk 9]
```diff
@@ -5,24 +5,46 @@ import (
...
 )
 
+var supportedOutputFormats = map[string]struct{}{
+	JSONFormat: {},
+	TEXTFormat: {},
+	MDFormat:   {},
+}
```
[the remark this answers is not in the document]
Comment (adisos) [reply cmd/analyzer/parse_args.go]: right, done

File: pkg/vpcmodel/cloudConfig.go (hunk 2 above)
Comment (adisos) [reply pkg/vpcmodel/cloudConfig.go]: ok, moved.

File: pkg/vpcmodel/subnetsConnectivity.go [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+			}
+		}
+	}
+	if subnetsConnectivityFromACLresources == nil {
+		return nil, errors.New("missing connectivity results from NACL resources")
+	}
+
+	pgwConns := map[string]ConfigBasedConnectivityResults{}
+	for _, r := range c.RoutingResources {
+		if r.Kind() == "PublicGateway" {
```
Comment (zivnevo) [inline pkg/vpcmodel/subnetsConnectivity.go:170]: Add const for the PublicGateway Kind

File: pkg/vpcmodel/subnetsConnectivity.go [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+	res += "Egress: \n"
+	for n, conn := range c.EgressAllowedConns {
+		res += subnetConnLine(n, conn)
+	}
+
+	return res
+}
+
+// print AllowedConns (not combined)
+func (v VPCsubnetConnectivity) printAllowedConns() {
```
Comment (zivnevo) [inline pkg/vpcmodel/subnetsConnectivity.go:54]: Having both pointer and non-pointer receivers for the same type is usually a bad idea

File: pkg/vpcmodel/subnetsConnectivity.go [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+			src := subnetNodeSet
+			dst := peerNode
+			combinedConns := conns
+
+			// peerNode kind is expected to be Subnet or External
+			peerNodeObj := v.cloudConfig.NameToResource[peerNode]
+			switch concPeerNode := peerNodeObj.(type) {
+			case NodeSet:
+				ingressConns := v.AllowedConns[concPeerNode.Name()].IngressAllowedConns[subnetNodeSet]
+				combinedConns = combinedConns.Intersection(ingressConns)
```
Comment (zivnevo) [inline pkg/vpcmodel/subnetsConnectivity.go:247]: Aren't we doing double work here? We should already have this <src, dst> pair in the map with the right ConnectionSet.
Can't we just `continue` if peerNode is not external?

File: pkg/vpcmodel/vpcConnectivity.go [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+func addDetailsLine(lines []string, details string) []string {
+	if details != "" {
+		lines = append(lines, details)
+	}
+	return lines
+}
+
+// add interface to output formatter
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/vpcConnectivity.go:105]: old comment.. deleted.

File: pkg/vpcmodel/vpcConnectivity.go [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+			dst := peerNode
+			combinedConns := conns
+			if peerNode.IsInternal() {
+				ingressConss := v.AllowedConns[peerNode].IngressAllowedConns[node]
+				combinedConns = combinedConns.Intersection(ingressConss)
+			}
+			if _, ok := v.AllowedConnsCombined[src]; !ok {
+				v.AllowedConnsCombined[src] = map[Node]*common.ConnectionSet{}
+			}
+			v.AllowedConnsCombined[src][dst] = combinedConns
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/vpcConnectivity.go]: done

File: pkg/vpcmodel/vpcConnectivity.go [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+	connResult := connMap[layer]
+	if isIngress {
+		return connResult.IngressAllowedConns[src]
+	}
+	return connResult.EgressAllowedConns[dst]
+}
+
+const (
+	// this layer is stateless, thus required in both directions for stateful connectivity computation
+	statefulRequiredLayerName = NaclLayer
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/vpcConnectivity.go:165]: ok, renaming

File: pkg/vpcmodel/subnetsConnectivity.go (hunk 11 above)
Comment (adisos) [reply pkg/vpcmodel/subnetsConnectivity.go:54]: right.. fixed

File: pkg/vpcmodel/vpcConnectivity.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply pkg/vpcmodel/vpcConnectivity.go]: `compteCombinedConnectionsPerDirection` -> `computeCombinedConnectionsPerDirection`

File: pkg/vpcmodel/subnetsConnectivity.go (hunk 10 above)
Comment (adisos) [reply pkg/vpcmodel/subnetsConnectivity.go:170]: done

File: pkg/vpcmodel/subnetsConnectivity.go (hunk 12 above)
Comment (adisos) [reply pkg/vpcmodel/subnetsConnectivity.go:247]: right, good catch! fixed here, and in `vpcConnectivity.go` as well

File: pkg/vpcmodel/subnetsConnectivity.go [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+
+	"github.com/np-guard/vpc-network-config-analyzer/pkg/common"
+)
+
+// VPCsubnetConnectivity captures allowed connectivity for subnets, considering nacl and pgw resources
+type VPCsubnetConnectivity struct {
+	// computed for each node (subnet), by iterating its ConnectivityResult for all relevant VPC resources that capture it
+	AllowedConns map[string]*ConfigBasedConnectivityResults
+	// combined connectivity - considering both ingress and egress per connection
+	AllowedConnsCombined map[string]map[string]*common.ConnectionSet
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/subnetsConnectivity.go:17]: I don't think so.. The `VPCConnectivity` struct is used for computation of connections between nodes, and the `VPCsubnetConnectivity` is used for such computation between subnets (it does not use `VPCConnectivity` object).

File: pkg/vpcmodel/vpcConnectivity.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/vpcConnectivity.go]: thanks, fixed

File: pkg/vpcmodel/subnetsConnectivity.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply pkg/vpcmodel/subnetsConnectivity.go:17]: ok, makes sense.

**Review by zivnevo — APPROVED** [review]
LGTM

**Pull request merged**